### ISSUE TITLE
Ansible ordering of hosts from inventory file changes between runs

### ISSUE THREAD

**Issue by psych0d0g** [ansible/ansible#4258]
i have a playbook that should run on only the first host of the ceph_mon group so i have a " when: inventory_hostname == groups.ceph_mon[0]" on each task but the ordering of the hosts variate between tasks.

Example Playbook:
ceph.yml: 

```
    ---
    - name: Collecting facts
      hosts: all
      gather_facts: yes

    - hosts: ceph_osd:ceph_mon
      name: get facts
      tasks:
    #    - action: ceph_facts
        - action: get_partitions.py

    - hosts: ceph_mon
      tasks:
        - include: ceph/tasks/mon_0.yml

    - hosts: ceph_mon
      tasks:
        - include: ceph/tasks/mon.yml

    - hosts: ceph_osd
      tasks:
        - include: ceph/tasks/osd.yml
```

mon_0.yml: 

```
    ---
    # To do only on the *first* mon client
    - name: Create key for client.admin and mon. on the first mon node
      shell: creates=/etc/ceph/ceph.keyring ceph-authtool --create-keyring --gen-key -n client.admin --set-uid=0 --cap mon 'allow *' --cap osd 'allow *' --cap mds 'allow *' /etc/ceph/ceph.keyring; ceph-authtool --gen-key -n mon. /etc/ceph/ceph.keyring --cap mon 'allow *'
      register: client_admin_key
      when: inventory_hostname == groups.ceph_mon[0]

    - name: Create monmap on the first mon node and store fsid in a dedicated file
      shell: monmaptool --create --add {{ inventory_hostname_short }} {{ hostvars[inventory_hostname_short]['ansible_eth2']['ipv4']['address'] }}:6789 /etc/ceph/monmap | grep fsid | awk {'print $4'} >> /etc/ceph/fsid
          creates=/etc/ceph/monmap
      when: inventory_hostname == groups.ceph_mon[0]

    - name: Create ceph working directory
      shell: mkdir /var/lib/ceph/mon/ceph-{{ inventory_hostname_short }}
          creates=/var/lib/ceph/mon/ceph-{{ inventory_hostname_short }}
      when: inventory_hostname == groups.ceph_mon[0]

    - name: Create local Ceph FileSystem
      command: ceph-mon --keyring /etc/ceph/ceph.keyring -i {{ inventory_hostname_short }} --mkfs --monmap /etc/ceph/monmap
          creates=/var/lib/ceph/mon/ceph-{{ inventory_hostname_short }}/store.db/LOCK
      when: inventory_hostname == groups.ceph_mon[0]

    - name: Touch something for the sake of touching it mwahaha
      shell: creates=/var/lib/ceph/mon/ceph-{{ inventory_hostname_short }}/sysvinit touch /var/lib/ceph/mon/ceph-{{ inventory_hostname_short }}/sysvinit
      when: inventory_hostname == groups.ceph_mon[0]

    - name: read fsid from file into global ansible var
      shell: cat /etc/ceph/fsid
      register: monmap
      when: inventory_hostname == groups.ceph_mon[0]

    - name: configure ceph with fsid from monmap
      template: src=ceph/templates/etc/ceph/ceph.conf.j2 dest=/etc/ceph/ceph.conf owner=root group=root mode=0644


    - name: Create key for OSD bootstrap
      shell: ceph-authtool --create-keyring --gen-key -n client.bootstrap-osd --cap mon 'allow profile bootstrap-osd' /etc/ceph/ceph.keyring.bootstrap-osd creates="/etc/ceph/ceph.keyring.bootstrap-osd"
      when: inventory_hostname == groups.ceph_mon[0]

    - name: Download the key from mon_0
      fetch: src=/etc/ceph/ceph.keyring dest=/tmp/fetched.ceph.keyring flat=yes fail_on_missing=yes
      when: inventory_hostname == groups.ceph_mon[0]

    - name: Download the OSD Bootstrap key from mon_0
      fetch: src=/etc/ceph/ceph.keyring.bootstrap-osd dest=/tmp/fetched.ceph.keyring.bootstrap-osd flat=yes fail_on_missing=yes
      when: inventory_hostname == groups.ceph_mon[0]

    - name: Download the monmap from mon_0
      fetch: src=/etc/ceph/monmap dest=/tmp/fetched.monmap flat=yes fail_on_missing=yes
      when: inventory_hostname == groups.ceph_mon[0]

    - name: Ensure ceph-mon is running
      service: name=ceph state=started pattern=/usr/bin/ceph-mon enabled=yes
      when: inventory_hostname == groups.ceph_mon[0]

    - name: feed our OSD bootstrap key to the cluster
      shell: ceph auth add client.bootstrap-osd -i /etc/ceph/ceph.keyring.bootstrap-osd
      when: inventory_hostname == groups.ceph_mon[0]

    - name: Create key for OSD bootstrap
      shell: ceph auth add client.bootstrap-osd -i /etc/ceph/ceph.keyring.bootstrap-osd
      when: inventory_hostname == groups.ceph_mon[0]

```

Inventory File:

```
    [14:36:39]cronix@cronix-work:~/ansible-dev/ceph_new$ cat /etc/ansible/hosts
    [ceph_mon]
    ceph-moni01
    ceph-moni02
    ceph-moni03

    [ceph_osd]
    ceph-osd01
    ceph-osd02
    ceph-osd03
    ceph-osd04
```

/etc/hosts:

```
    14:36:28]cronix@cronix-work:~/ansible-dev/ceph_new$ cat /etc/hosts
    127.0.0.1 localhost
    127.0.1.1 ********************

    192.168.56.11 ceph-moni01
    192.168.56.12 ceph-moni02
    192.168.56.13 ceph-moni03
    192.168.56.21 ceph-osd01
    192.168.56.22 ceph-osd02
    192.168.56.23 ceph-osd03
    192.168.56.24 ceph-osd04
```

Output: http://pastebin.com/t3vcfPca
It should only run on ceph_moni01 but it runs on 02 or 03 aswell sometimes.

**Comment by mpdehaan** [comment]
notes:

groups are defined like

[foo]
a
b
c
d

and when: inventory_hostname == groups.foo[0]

groups.foo[0] is not consistently sorted between execution runs.

This should be preserved in the order listed in the inventory file, not hash order, and is a relatively important bug that we should fix soon.

**Comment by mpdehaan** [comment]
Thanks for update!

**Comment by mpdehaan** [comment]
BTW, an easier way to specify the include:
- include: foo.yml
  when: x == 2

(This avoids needing to repeat the conditional for each line in the file, which makes the other playbook a lot shorter.  It doesn't fix this above problem)

**Comment by psych0d0g** [comment]
thanks for the hint, that will be changed as soon as the ordering is fixed, i depend on that and i cant continue until that works, i think ill check earlier versions of ansible so i can see when that bug was introduced

**Comment by psych0d0g** [comment]
it seems that the bug was introduced with ansible 1.2.3, 1.2.2 is the last version that works as expected here

**Comment by mpdehaan** [comment]
This has been fixed on the development branch in 8fc46a3a5a286bee5e429b921ef8af9ea7ecbc31.

Thanks!
